util/persistence.py

```diff
@@ -0,0 +1,242 @@
+        # finally: saving the new checkpoint
+        self.saver.save(session, filename, write_state=False)
+
+    def _init(self, session):
+        '''
+        Initializes the session.
+        '''
+        log.info('No checkpoint to load - initializing new model...')
+        session.run(self.init_op)
+
+    def _restore(self, session, filename):
+        '''
+        Restores a checkpoint file-set.
+        '''
+        log.info('Restoring checkpoint "%s"...' % filename)
+        self.saver.restore(session, filename)
+
+    def start(self, session):
+        '''
+        Starts a training session by either restoring the right checkpoint
+        or by initializing the model.
+        Should be called at the beginning of a training session.
+        '''
+        if not self.checkpoint_dir:
+            self._init(session)
+            return
+        checkpoint = None
+        # collect available checkpoints
+        checkpoints = self._get_checkpoints()
+        if self.load == 'recent':
+            if len(checkpoints) > 0:
+                # pick the most recent checkpoint (by means of global-index),
+                # no matter if this is an epoch or intermediate one
+                checkpoint = checkpoints[-1]
+        else:
+            # epoch checkpoints only, ordered by epoch number
+            checkpoints = sorted([cp for cp in checkpoints if cp[0] > 0], key=lambda cp: cp[0])
+            if len(checkpoints) > 0:
+                if self.load == 'best-dev':
+                    # get epoch numbers of existing checkpoints
+                    available = [cp[0] for cp in checkpoints]
+                    # and intersect them with result-file-listed epochs featuring a dev-loss
+                    available = [r for r in self._results if r[0] in available and not r[2] is None]
+                    # if there are any,
+                    if len(available) > 0:
+                        # pick the one with dev-loss minimum
+                        epoch = min(available, key=lambda r: r[2])[0]
+                        checkpoint = next(cp for cp in checkpoints if cp[0] == epoch)
+                elif self.load == 'last-epoch':
+                    # pick the one with the highest epoch number
+                    checkpoint = checkpoints[-1]
+        if checkpoint:
+            # so we got a checkpoint file-set - restoring it...
+            self._restore(session, self._get_filename(checkpoint[2]))
+        elif self.load != 'init' and len(glob.glob(self.load + '*')) > 0:
+            # if self.load is not a keyword, but a path-prefix to existing files, we restore it...
+            self._restore(session, self.load)
+        else:
+            # final fallback -> fresh start (also self.load == 'init')
+            self._init(session)
+        # setting t0 for intermediate checkpointing
+        self._t0 = time.time()
+        # return epoch history
+        return self._results
+
+    def step(self, session, global_index):
```

Comment:
Maybe there's value in some `log.debug()` when we get called, when we do nothing, and when we call `_prune_and_save()` ?

Reply:
This will be far too much noise on th log for just logging "I'm not doing anything". `_prune_and_save()` logs by itself.